phalcon/Storage/Adapter/Weak.zep

```diff
@@ -0,0 +1,235 @@
+     * @throws SupportException
+     */
+    public function __construct(<SerializerFactory> factory, array! options = [])
+    {
+        let this->defaultSerializer = "none",
+            this->prefix                 = "",
+            this->options                = options;
+    }
+
+     /**
+     * Flushes/clears the cache
+     */
+    public function clear() -> bool
+    {
+        let this->weakList = [];
+        return true;
+    }
+
+    /**
+     * Decrements a stored number
+     *
+     * @param string $key
+     * @param int    $value
+     *
+     * @return bool|int
+     */
+    public function decrement(string! key, int value = 1) -> int | bool
+    {
+        return false;
+    }
+
+    /**
+     * Deletes data from the adapter
+     *
+     * @param string $key
+     *
+     * @return bool
+     */
+    public function delete(string! key) -> bool
+    {
+        var exists;
+        if key === this->fetching {
+            return false;
+        }
+        let exists = isset this->weakList[key];
+        unset(this->weakList[key]);
+        return exists;
+    }
+
+    /**
+     * Stores data in the adapter
+     *
+     * @param string $prefix
+     *
+     * @return array
+     */
+    public function getKeys(string prefix = "") -> array
+    {
+        var keys, key;
+        array results;
+        let keys = array_keys(this->weakList);
+        if unlikely "" !== prefix {
+            let results = [];
+            for key in keys {
+                if starts_with(key, prefix) {
+                    let results[] = key;
+                }
+            }
+            return results;
+        }
+        return keys;
+    }
+
+    /**
+     * Increments a stored number
+     *
+     * @param string $key
+     * @param int    $value
+     *
+     * @return bool|int
+     */
+    public function increment(string! key, int value = 1) -> int | bool
+    {
+        return false;
+    }
+
+    /**
+    * Reads data from the adapter
+    *
+    * @param string key
+    * @param mixed|null   defaultValue
+    *
+    * @return mixed
+    */
+    public function get(string! key, var defaultValue = null) -> var
+    {
+        var value, wr;
+        /**
+         * while getting a key, garbage collection might be triggered,
+         * this will stop unsetting the key, will not stop however the model gets destroid by GC,
+         * this is for the destruct that is in the model
+         * not do destroy the key before getting it.
+         */
+        let this->fetching = key;
+        if false === isset this->weakList[key] {
+            let this->fetching = null;
+            return defaultValue;
+        }
+        let wr = this->weakList[key];
+        let value = wr->get();
+        let this->fetching = null;
+        /**
+         * value could be null, object could be destroid while fetching
+         */
+        if null === value {
+            this->delete(key);
+        }
+        return value;
+    }
+
```

Comment:
I don't think the `doGet` is required here. The `get()` in the abstract class checks if the key exists then if not returns the default value and then it performs the `doGet()`

So I was wondering if the above (get) can be adjusted so that the code is used from the abstract class and when the actual `get` happens, that goes in the `doGet()` method (including the `delete() part that is in line 162

Reply:
Maybe it's not required, Weak does not have an Adapter, so this->adapter in the AbstractAdapter doGet() wont work.
I checked this against Storage\Adapter\Memory that currently is like this.

```
    protected function doGet(string key)
    {
        return this->data[this->getPrefixedKey(key)];
    }

```

I think it's like this mainly for compatibility reasons.
maybe I can scrap it.

Has for the get(), my reasoning is this is a bit optimized, I could scrape it and use the AbstractAdapter::get(), make the get() into the doGet() and set the serializer to null, just trying to avoid a jump to has() and another jump to getUnserializedData() just to return content unaltered.
Trying to save 2 or 3 jumps, in my use case is going to use Weak every time ResultSet::current() is called.

I could do something more optimized in the model later on.

What do you think?

Reply:
If we are to override `get` and have it run from the Weak, then no need for the `doGet`. 

I would go with that really since the abstract `get` will never be called

Reply:
OK, if you are ok with this, I will remove the doGet().
Thanks!